File: 3. Arrays/numSubarrayProductLessThanK.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 713. Subarray Product Less Than K
 */
var numSubarrayProductLessThanK = function(nums, k) {
    if (k <= 1) return 0
    let start = -1;
    let output = 0;
    let totalProduct = 1;
    for (let i = 0; i < nums.length; i++) {
        totalProduct *= nums[i]
        while (start < i && totalProduct >= k) {
            start++;
            totalProduct /= nums[start]
        }
        output += i - start
    }

    return output
};

File: 3. Arrays/num_subarray_product_less_than_k.py
class Solution:
    def numSubarrayProductLessThanK(self, nums: List[int], k: int) -> int:
        output = 0
        start = -1
        totalProduct = 1
        for idx in range(len(nums)):
            totalProduct *= nums[idx]
            while (start < idx and totalProduct >= k):
                start += 1
                totalProduct /= nums[start]

            output += idx - start
        return output
